_posts/2022-04-24-ui-architecture.md

```diff
@@ -0,0 +1,264 @@
+
+Ron Minsky has [stated][Signals and Threads: Building a UI framework] "hidden inside of every UI framework is some kind of incrementalization framework." Xilem unapologetically contains at its core a lightweight change propagation engine, similar in scope to the attribute graph of SwiftUI, but highly specialized to the needs of UI, and in particular with a lightweight approach to *downward* propagation of dependencies, what in React would be stated as the flow of props into components.
+
+In this particular case, that incremental change propagation is best represented as a *memoization* node, yet another implementation of the View trait. A memoization node takes a data value (which supports both `Clone` and equality testing) and a closure which accepts that same data type. On rebuild, it compares the data value with the previous version, and only runs the closure if it has changed. The signature of this node is very similar to [Html.Lazy] in Elm.
+
+Comparing a number is extremely cheap (especially because all this happens with static typing, so no boxing or downcasting is needed), but the cost of equality comparison is a valid concern for larger, aggregate data structures. Here, immutable data structures (adapted from the existing Druid architecture) can work very well.
+
+Let's say there's a parent object that contains all the app state, including a sizable child component. The type would look something like this:
+
+```rust
+#[derive(Clone)]
+struct Parent {
+    stuff: Stuff,
+    child: Arc<Child>,
+}
+```
+
+And at the top of the tree we can use a memoize node with type `Arc<Parent>`, and the equality comparison [pointer equality] on the `Arc` rather than a deep traversal into the structure (as might be the case with a derived `PartialEq` impl). The child component attaches with both a Memoize and an Adapt node.
+
+The details of the Adapt node are interesting. Here's a simple approach:
+
+```rust
+adapt(
+    |data: &mut Arc<Parent>, thunk| thunk.call(&mut Arc::make_mut(data).child),
+    child_view(...) // which has Arc<Child> as its app data type
+)
+```
+
+Whenever events propagate into the child, `make_mut` creates a copy of the parent struct, which will then not be pointer-equal to the version stored in the memoize node. If such events are relatively rare, or if they nearly always end up mutating the child state, then this approach is reasonable. However, it is possible to be even finer grain:
+
+```rust
+adapt(
+    |data: &mut Arc<Parent>, thunk| {
+        let mut child = data.child.clone();
+        thunk.call(&mut child),
+        if !Arc::ptr_eq(&data.child, &child) {
+            Arc::make_mut(data).child = child;
+        }
+    },
+    child_view(...) // which has Arc<Child> as its app data type
+)
+```
+
+This logic propagates the change up the tree *only if* the child state has actually changed.
```

Comment:
Don't you want change to propagate down the tree? Actually, which tree is this referring to?

Reply:
The app state tree. But I'm going to change the way I word this, because there's no *requirement* that the app state be modeled as a tree, unlike original Druid. It's quite fine for the app state to be a graph (and we're going to have this in Runebender, as glyphs referencing components is definitely graph rather than tree structure). Good catch, thanks.